internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,511 @@
+		Subject: pkix.Name{
+			Organization: []string{"Canonical Ltd."},
+		},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+	// DER encoded bytes
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, &privateKey.PublicKey, privateKey)
+	if err != nil {
+		return nil, err
+	}
+	// Load a clean X509 certificate from the DER encoded bytes.
+	certificate, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		return nil, err
+	}
+	keypair := &TLSKeyPair{privateKey: privateKey, Certificate: certificate}
+	keypair.addFingerprint()
+	return keypair, nil
+}
+
+// persistTLSKeyPair persists the TLS keypair in a single PEM file. The file name
+// is the certificate fingerprint string, with a .pem suffix. See the
+// addFingerprint method for information on how the fingerprint is generated.
+func (m *TLSManager) persistTLSKeyPair(keypair *TLSKeyPair) (err error) {
+	// If the TLS directory does not yet exist, create it.
+	_, err = os.Stat(m.tlsDir)
+	if os.IsNotExist(err) {
+		// Create the leaf directory node with 0700 permissions.
+		err = os.Mkdir(m.tlsDir, 0700)
+	}
+	if err != nil {
+		return fmt.Errorf("cannot create directory leaf of path %v: %w", m.tlsDir, err)
+	}
+
+	pemPath := filepath.Join(m.tlsDir, fmt.Sprintf("%v.pem", keypair.Fingerprint()))
+	pemFile, err := os.OpenFile(pemPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errors.Join(err, pemFile.Sync())
+		err = errors.Join(err, pemFile.Close())
+	}()
+
+	pemData, err := keypair.toPEMBytes()
+	if err != nil {
+		return err
+	}
+	if _, err = pemFile.Write(pemData); err != nil {
+		return err
+	}
+	return nil
+}
+
+// TLSKeyPair represents a TLS key pair (see tls.LoadX509KeyPair), which
+// includes both the x509 certificate and the private key (unexported). This
+// type can directly be converted to a tls.Certificate for configuring a TLS
+// listener, or exported as a PEM encoded x509 certificate for easy transport
+// over the daemon API.
+type TLSKeyPair struct {
+	privateKey  any
+	Certificate *x509.Certificate
+	fingerprint string
+}
+
+// Compare compares two keypairs and returns an int as used by sorting
+// functions.
+func (t TLSKeyPair) Compare(other *TLSKeyPair) int {
+	// Start time first.
+	notBeforeSort := t.Certificate.NotBefore.Compare(other.Certificate.NotBefore)
+	if notBeforeSort != 0 {
+		return notBeforeSort
+	}
+	// End time next.
+	notAfterSort := t.Certificate.NotAfter.Compare(other.Certificate.NotAfter)
+	if notAfterSort != 0 {
+		return notAfterSort
+	}
+	// Fingerprint last.
+	return strings.Compare(t.Fingerprint(), other.Fingerprint())
+}
+
+// Fingerprint returns an unpadded base32 encoded SHA512/384 hash of the raw
+// x509 certificate.
+func (t TLSKeyPair) Fingerprint() string {
+	return t.fingerprint
```

Comment:
You could simplify your life and just calculate the fingerprint right here, though this might have a performance impact. Alternatively, you could calculate it lazily (i.e. leave it `nil` until someone calls this function, and only calculate if the field is nil), combining simplicity and performance.

Reply:
PR reworked. N/A for current PR. Resolving.